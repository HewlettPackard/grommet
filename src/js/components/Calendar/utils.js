
// Utility functions for the Calendar.
// Just what's needed to avoid having to include a dependency like momentjs.

const DAY_MILLISECONDS = 24 * 60 * 60 * 1000;

export const addDays = (date, days) =>
  (new Date(date.getTime() + (DAY_MILLISECONDS * days)));

export const subtractDays = (date, days) => addDays(date, -days);

export const addMonths = (date, months) => {
  const result = new Date(date);
  const years = Math.floor((date.getMonth() + months) / 12);
  result.setFullYear(date.getFullYear() + years);
  result.setMonth((date.getMonth() + months) % 12);
  return result;
};

export const subtractMonths = (date, months) => addMonths(date, -months);

export const sameDay = (date1, date2) => (
  date1.getFullYear() === date2.getFullYear() &&
  date1.getMonth() === date2.getMonth() &&
  date1.getDate() === date2.getDate()
);

export const sameDayOrAfter = (date1, date2) =>
  (date1.getFullYear() > date2.getFullYear() ||
    (date1.getFullYear() === date2.getFullYear() && (
      date1.getMonth() > date2.getMonth() || (
        date1.getMonth() === date2.getMonth() && (date1.getDate() >= date2.getDate())
      )
    ))
  );

export const sameDayOrBefore = (date1, date2) =>
  (date1.getFullYear() < date2.getFullYear() ||
    (date1.getFullYear() === date2.getFullYear() && (
      date1.getMonth() < date2.getMonth() || (
        date1.getMonth() === date2.getMonth() && (date1.getDate() <= date2.getDate())
      )
    ))
  );

export const daysApart = (date1, date2) =>
  Math.floor((date1.getTime() - date2.getTime()) / DAY_MILLISECONDS);

// betweenDates takes and array of two elements and checks if the
// supplied date lies between them, inclusive.
// returns 2 if exact match to one end, 1 if between, undefined otherwise
export const betweenDates = (date, dates) => {
  let result;
  if (dates) {
    const [from, to] = dates.map(d => new Date(d));
    if (sameDay(date, from) || sameDay(date, to)) {
      result = 2;
    } else if (sameDayOrAfter(date, from) && sameDayOrBefore(date, to)) {
      result = 1;
    }
  } else {
    result = 1;
  }
  return result;
};

// withinDates takes and array of string dates or 2 element arrays and
// checks whether the supplied date matches any string or is between
// any dates in arrays.
// returns 2 if exact match, 1 if between, undefined otherwise
export const withinDates = (date, dates) => {
  let result;
  if (dates) {
    if (Array.isArray(dates)) {
      dates.some((d) => {
        if (typeof d === 'string') {
          if (sameDay(date, new Date(d))) {
            result = 2;
          }
        } else {
          result = betweenDates(date, d);
        }
        return result;
      });
    } else if (sameDay(date, new Date(dates))) {
      result = 2;
    }
  }
  return result;
};
