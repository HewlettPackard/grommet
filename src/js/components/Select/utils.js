import React, { forwardRef } from 'react';
import styled from 'styled-components';
import { Box } from '../Box';
import { Button } from '../Button';
import { DropButton } from '../DropButton';
import { Text } from '../Text';
import { TextInput } from '../TextInput';
import {
  getHoverIndicatorStyle,
  selectedStyle,
  controlBorderStyle,
  normalizeColor,
} from '../../utils';

export const applyKey = (option, key) => {
  if (option === undefined) return undefined;
  if (typeof key === 'object') return applyKey(option, key.key);
  if (typeof key === 'function') return key(option);
  if (key !== undefined) return option[key];
  return option;
};

// position relative is so scroll can be managed correctly
export const OptionsBox = styled.div`
  position: relative;
  scroll-behavior: smooth;
  overflow: auto;
  outline: none;
`;

export const HiddenInput = styled.input`
  display: none;
`;

export const SelectOption = styled(Button)`
  ${(props) => props.selected && props.textComponent && selectedStyle}
  // applies theme.global.hover.background to the active
  // option for mouse and keyboard interactions
  ${(props) =>
    props.active &&
    getHoverIndicatorStyle(
      !props.children && !props.theme.select.options ? undefined : 'background',
      props.theme,
    )}
  display: block;
  width: 100%;
`;

export const SelectTextInput = styled(TextInput)`
  cursor: ${(props) => (props.defaultCursor ? 'default' : 'pointer')};
`;

export const StyledSelectDropButton = styled(DropButton)`
  ${(props) => !props.callerPlain && controlBorderStyle};
  ${(props) =>
    props.theme.select &&
    props.theme.select?.control &&
    props.theme.select?.control?.extend};
  ${(props) => props.open && props.theme.select?.control?.open};
`;

export const DefaultSelectTextInput = forwardRef(
  (
    { a11yTitle, disabled, id, name, placeholder, value, size, theme, ...rest },
    ref,
  ) => (
    <SelectTextInput
      a11yTitle={a11yTitle}
      // When Select is disabled, we want to show a default cursor
      // but not have disabled styling come from TextInput
      // Disabled can be a bool or an array of options to disable.
      // We only want to disable the TextInput if the control
      // button should be disabled which occurs when disabled
      // equals true.
      defaultCursor={disabled === true || undefined}
      focusIndicator={false}
      id={id ? `${id}__input` : undefined}
      name={name}
      ref={ref}
      {...rest}
      tabIndex="-1"
      type="text"
      placeholder={placeholder}
      plain
      readOnly
      value={value}
      size={size}
      theme={theme}
    />
  ),
);

export const EmptySearchOption = ({
  emptySearchMessage,
  selectOptionsStyle,
  theme,
}) => (
  <SelectOption
    key="search_empty"
    tabIndex="0"
    role="menuitem"
    hoverIndicator="background"
    disabled
    aria-live="polite"
  >
    <Box {...selectOptionsStyle}>
      <Text {...theme.select.container.text}>{emptySearchMessage}</Text>
    </Box>
  </SelectOption>
);

export const getOptionLabel = (index, options, labelKey) =>
  applyKey(options[index], labelKey);

export const getOptionValue = (index, options, valueKey) =>
  applyKey(options[index], valueKey);

export const checkDisabled = (
  index,
  disabled,
  disabledKey,
  options,
  valueKey,
) => {
  const option = options[index];
  let result;
  if (disabledKey) {
    result = applyKey(option, disabledKey);
  } else if (Array.isArray(disabled)) {
    if (typeof disabled[0] === 'number') {
      result = disabled.indexOf(index) !== -1;
    } else {
      const optionVal = getOptionValue(index, options, valueKey);
      result = disabled.indexOf(optionVal) !== -1;
    }
  }
  return result;
};

export const calcValuedValue = (value, valueKey) => {
  if (Array.isArray(value))
    return value.map((v) =>
      valueKey && valueKey.reduce ? v : applyKey(v, valueKey),
    );
  return valueKey && valueKey.reduce ? value : applyKey(value, valueKey);
};

export const changeEvent = (inputRef, nextValue) => {
  // Calling set value function directly on input because React library
  // overrides setter `event.target.value =` and loses original event
  // target fidelity.
  // https://stackoverflow.com/a/46012210
  const nativeInputValueSetter = Object.getOwnPropertyDescriptor(
    window.HTMLInputElement.prototype,
    'value',
  ).set;
  nativeInputValueSetter.call(inputRef.current, nextValue);
  const event = new Event('input', { bubbles: true });
  inputRef.current.dispatchEvent(event);
};

export const getSelectIcon = (icon, theme, open) => {
  let SelectIcon;
  switch (icon) {
    case false:
      break;
    case true:
    case undefined:
      SelectIcon =
        open && theme.select.icons.up
          ? theme.select.icons.up
          : theme.select.icons.down;
      break;
    default:
      SelectIcon = icon;
  }
  return SelectIcon;
};

// if labelKey is a function and valueLabel is not defined
// we should use the labelKey function to display the
// selected value
export const getDisplayLabelKey = (
  labelKey,
  allOptions,
  optionIndexesInValue,
  selectValue,
) => {
  const optionLabelKey = applyKey(
    allOptions[optionIndexesInValue[0]],
    labelKey,
  );
  if (
    !selectValue &&
    optionIndexesInValue.length === 1 &&
    typeof optionLabelKey === 'object'
  )
    return optionLabelKey;
  return undefined;
};

export const getIconColor = (theme) =>
  normalizeColor(theme.select.icons.color || 'control', theme);
