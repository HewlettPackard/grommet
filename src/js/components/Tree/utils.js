export const setupNodes = nodes =>
  nodes.map(node => ({
    name: node.name,
    childs: setupNodes(node.childs || []),
    opened: false,
  }));

export const isSameNode = (node1, node2) => node1 === node2;

export const changeNode = (positions, nodes) => {
  const [currentPosition, ...otherPositions] = positions;
  const currentNode = nodes[currentPosition];

  if (otherPositions.length) {
    return nodes.map(node =>
      isSameNode(node, currentNode)
        ? {
            ...node,
            childs: changeNode(otherPositions, node.childs),
          }
        : { ...node },
    );
  }

  return nodes.map(node =>
    isSameNode(node, currentNode)
      ? {
          ...node,
          opened: !node.opened,
        }
      : { ...node },
  );
};
