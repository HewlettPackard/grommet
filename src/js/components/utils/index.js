export * from './styles';
export * from './mixins';
export * from './DOM';
export { default as Drop } from './Drop';
