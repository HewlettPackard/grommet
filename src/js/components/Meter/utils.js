import { colorForName } from '../../utils';

export const strokeProps = (color, theme) => {
  const result = {};
  if (color) {
    if (typeof color === 'object') {
      result.stroke = colorForName(color.color, theme);
      if (color.opacity) {
        result.strokeOpacity = (color.opacity === true ?
          theme.global.opacity.medium : theme.global.opacity[color.opacity]);
      }
    } else {
      result.stroke = colorForName(color, theme);
    }
  }
  return result;
};

export default { strokeProps };
