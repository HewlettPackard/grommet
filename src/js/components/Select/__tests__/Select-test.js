import React from 'react';
import 'jest-styled-components';
import renderer from 'react-test-renderer';
import { cleanup, render, renderIntoDocument, Simulate } from 'react-testing-library';

import { createPortal, expectPortal } from '../../../utils/portal';

import { Select } from '../';

describe('Select', () => {
  beforeEach(createPortal);

  afterEach(cleanup);

  test('basic', () => {
    const component = renderer.create(
      <Select id='test-select' options={['one', 'two']} />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('opens', (done) => {
    window.scrollTo = jest.fn();
    const { getByTestId, container } = renderIntoDocument(
      <Select data-testid='test-select' id='test-select' options={['one', 'two']} />
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(document.getElementById('test-select__drop')).toBeNull();

    Simulate.click(getByTestId('test-select'));

    expect(container.firstChild).toMatchSnapshot();
    expectPortal('test-select__drop').toMatchSnapshot();

    setTimeout(() => {
      expect(document.activeElement).toMatchSnapshot();
      done();
    }, 100);
  });

  test('complex options and children', () => {
    const { getByTestId, container } = renderIntoDocument(
      <Select
        data-testid='test-select'
        id='test-select'
        options={[{ test: 'one' }, { test: 'two' }]}
      >
        {option => <span>{option.test}</span>}
      </Select>
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(document.getElementById('test-select__drop')).toBeNull();

    Simulate.click(getByTestId('test-select'));

    expect(container.firstChild).toMatchSnapshot();
    expectPortal('test-select__drop').toMatchSnapshot();
  });

  test('search', () => {
    jest.useFakeTimers();
    const onSearch = jest.fn();
    const { getByTestId, container } = renderIntoDocument(
      <Select
        data-testid='test-select'
        id='test-select'
        options={['one', 'two']}
        onSearch={onSearch}
      />
    );
    expect(container.firstChild).toMatchSnapshot();

    Simulate.click(getByTestId('test-select'));

    expectPortal('test-select__drop').toMatchSnapshot();

    setTimeout(() => {
      jest.runAllTimers();

      expect(document.activeElement).toMatchSnapshot();

      document.activeElement.value = 'a';
      Simulate.input(document.activeElement);

      expect(onSearch).toBeCalledWith('a');
    }, 200);
  });

  // NOTE: This isn't really a test for Select
  // test('closes drop on esc', () => {
  //   const onClose = jest.fn();
  //   const component = mount(
  //     <Select id='test-select' options={['one', 'two']} onClose={onClose} />
  //   );
  //
  //   Simulate.keyDown(
  //     component.getDOMNode(),
  //     { key: 'Down', keyCode: 40, which: 40 }
  //   );
  //
  //   expectPortal('test-select__drop').toMatchSnapshot();
  //   expect(component.getDOMNode()).toMatchSnapshot();
  //
  //   Simulate.keyDown(
  //     document.getElementById('test-select__drop'),
  //     { key: 'Esc', keyCode: 27, which: 27 }
  //   );
  //
  //   expect(onClose).toBeCalled();
  //   expect(document.getElementById('test-select__drop')).toBeNull();
  //   expect(component.getDOMNode()).toMatchSnapshot();
  // });

  test('select an option', () => {
    window.scrollTo = jest.fn();
    const onChange = jest.fn();
    const { getByTestId, container } = renderIntoDocument(
      <Select
        data-testid='test-select'
        id='test-select'
        options={['one', 'two']}
        onChange={onChange}
      />
    );
    expect(container.firstChild).toMatchSnapshot();

    Simulate.click(getByTestId('test-select'));

    // pressing enter here nothing will happen
    Simulate.click(
      document.getElementById('test-select__drop').querySelector('button')
    );
    expect(onChange).toBeCalled();
    expect(window.scrollTo).toBeCalled();
  });

  test('select an option with complex options', () => {
    window.scrollTo = jest.fn();
    const onChange = jest.fn();
    const { getByTestId, container } = renderIntoDocument(
      <Select
        data-testid='test-select'
        id='test-select'
        plain={true}
        value={<span>one</span>}
        options={[{ test: 'one' }, { test: 'two' }]}
        onChange={onChange}
      >
        {option => <span>{option.test}</span>}
      </Select>
    );
    expect(container.firstChild).toMatchSnapshot();

    Simulate.click(getByTestId('test-select'));

    // pressing enter here nothing will happen
    Simulate.click(
      document.getElementById('test-select__drop').querySelector('button')
    );
    expect(onChange).toBeCalled();
    expect(window.scrollTo).toBeCalled();
  });

  test('select an option with enter', () => {
    window.scrollTo = jest.fn();
    const onChange = jest.fn();
    const { getByTestId, container } = renderIntoDocument(
      <Select
        data-testid='test-select'
        id='test-select'
        options={['one', 'two']}
        onChange={onChange}
      />
    );
    expect(container.firstChild).toMatchSnapshot();

    Simulate.click(getByTestId('test-select'));

    const preventDefault = jest.fn();
    Simulate.keyDown(
      document.getElementById('test-select__select-drop'),
      { key: 'Down', keyCode: 40, which: 40, preventDefault },
    );
    expect(preventDefault).toBeCalled();
    Simulate.keyDown(
      document.getElementById('test-select__select-drop'),
      { key: 'Up', keyCode: 38, which: 38, preventDefault },
    );
    expect(preventDefault).toBeCalled();
    Simulate.keyDown(
      document.getElementById('test-select__select-drop'),
      { key: 'Enter', keyCode: 13, which: 13, preventDefault },
    );
    expect(preventDefault).toBeCalled();
    expect(onChange).toBeCalled();
    expect(window.scrollTo).toBeCalled();
  });

  test('size', () => {
    const component = renderer.create(
      <Select
        id='test-select'
        size='large'
        options={['one', 'two']}
        selected={[]}
        value={[]}
      />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('multiple', () => {
    const component = renderer.create(
      <Select
        id='test-select'
        multiple={true}
        options={['one', 'two']}
        selected={[]}
        value={[]}
      />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('multiple values', () => {
    const { getByTestId, container } = render(
      <Select
        data-testid='test-select'
        id='test-select'
        multiple={true}
        options={['one', 'two']}
        selected={[0, 1]}
        value={['one', 'two']}
      />
    );
    expect(container.firstChild).toMatchSnapshot();

    Simulate.click(getByTestId('test-select'));

    expect(container.firstChild).toMatchSnapshot();
    expectPortal('test-select__drop').toMatchSnapshot();
  });

  test('select another option', () => {
    const onChange = jest.fn();
    const { getByTestId, container } = renderIntoDocument(
      <Select
        data-testid='test-select'
        id='test-select'
        multiple={true}
        options={['one', 'two']}
        onChange={onChange}
        value={['two']}
        selected={[1]}
      />
    );
    expect(container.firstChild).toMatchSnapshot();

    Simulate.click(getByTestId('test-select'));

    Simulate.click(
      document.getElementById('test-select__drop').querySelector('button')
    );
    expect(onChange).toBeCalled();
  });

  test('deselect an option', () => {
    const onChange = jest.fn();
    const { getByTestId, container } = renderIntoDocument(
      <Select
        data-testid='test-select'
        id='test-select'
        multiple={true}
        options={['one', 'two']}
        onChange={onChange}
        value={['one']}
        selected={[0]}
      />
    );
    expect(container.firstChild).toMatchSnapshot();

    Simulate.click(getByTestId('test-select'));

    Simulate.click(
      document.getElementById('test-select__drop').querySelector('button')
    );
    expect(onChange).toBeCalled();
  });

  test('disabled', () => {
    const { getByTestId, container } = renderIntoDocument(
      <Select
        data-testid='test-select'
        id='test-select'
        disabled={true}
        options={['one', 'two']}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(document.getElementById('test-select__drop')).toBeNull();

    Simulate.click(getByTestId('test-select'));

    expect(container.firstChild).toMatchSnapshot();
    expect(document.getElementById('test-select__drop')).toBeNull();
  });
});
