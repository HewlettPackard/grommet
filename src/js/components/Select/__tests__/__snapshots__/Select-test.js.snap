// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select closes drop on esc 1`] = `
<div
  class="StyledBox-bStriS iQipYn"
  direction="column"
  id="select-drop__select"
>
  <div
    class="StyledBox-bStriS feBFzz"
    direction="column"
    overflow="auto"
  >
    <div
      class="StyledBox-bStriS jDxevM"
      direction="column"
      role="menubar"
      tabindex="-1"
    >
      <button
        class="StyledButton-gXzblK khQPBH"
        role="menuitem"
        type="button"
      >
        <div
          class="StyledBox-bStriS gedxom"
          direction="column"
        >
          <span
            class="StyledText-eJMSZI bUVmvh"
          >
            one
          </span>
        </div>
      </button>
      <button
        class="StyledButton-gXzblK khQPBH"
        role="menuitem"
        type="button"
      >
        <div
          class="StyledBox-bStriS gedxom"
          direction="column"
        >
          <span
            class="StyledText-eJMSZI bUVmvh"
          >
            two
          </span>
        </div>
      </button>
    </div>
  </div>
</div>
`;

exports[`Select closes drop on esc 2`] = `
"

.iQipYn {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 0;
  padding: 0;
}

@media only screen and (min-width:700px) {
  .cfJbKA {
    -webkit-transition: 0.1s ease-in-out;
    transition: 0.1s ease-in-out;
  }
}

@media only screen and (min-width:700px) {
  .khQPBH {
    -webkit-transition: 0.1s ease-in-out;
    transition: 0.1s ease-in-out;
  }
}"
`;

exports[`Select closes drop on esc 3`] = `
<button
  class="StyledButton-gXzblK cfJbKA"
  aria-label="undefined"
  type="button"
>
  <div
    class="StyledBox-bStriS janhJy"
    direction="row"
    aria-hidden="true"
  >
    <div
      class="StyledTextInput__StyledTextInputContainer-bPTWQl hasTun"
    >
      <input
        type="text"
        class="StyledTextInput-bzOzsW VzRqL"
        id="select"
        autocomplete="off"
        style="cursor: pointer;"
        tabindex="-1"
        readonly=""
      />
    </div>
    <div
      class="StyledBox-bStriS jagiFv"
      direction="column"
      style="min-width: auto;"
    >
      <svg
        class="StyledIcon-gTfjEw htxeRg"
        width="24px"
        height="24px"
        viewBox="0 0 24 24"
        version="1.1"
        role="img"
        aria-label="FormDown"
      >
        <polyline
          fill="none"
          stroke="#000"
          stroke-width="2"
          points="18 9 12 15 6 9"
        />
      </svg>
    </div>
  </div>
</button>
`;

exports[`Select closes drop on esc 4`] = `
<button
  class="StyledButton-gXzblK cfJbKA"
  aria-label="undefined"
  type="button"
>
  <div
    class="StyledBox-bStriS janhJy"
    direction="row"
    aria-hidden="true"
  >
    <div
      class="StyledTextInput__StyledTextInputContainer-bPTWQl hasTun"
    >
      <input
        type="text"
        class="StyledTextInput-bzOzsW VzRqL"
        id="select"
        autocomplete="off"
        style="cursor: pointer;"
        tabindex="-1"
        readonly=""
      />
    </div>
    <div
      class="StyledBox-bStriS jagiFv"
      direction="column"
      style="min-width: auto;"
    >
      <svg
        class="StyledIcon-gTfjEw htxeRg"
        width="24px"
        height="24px"
        viewBox="0 0 24 24"
        version="1.1"
        role="img"
        aria-label="FormDown"
      >
        <polyline
          fill="none"
          stroke="#000"
          stroke-width="2"
          points="18 9 12 15 6 9"
        />
      </svg>
    </div>
  </div>
</button>
`;

exports[`Select mounts 1`] = `
<button
  class="StyledButton-gXzblK cfJbKA"
  aria-label="undefined"
  type="button"
>
  <div
    class="StyledBox-bStriS janhJy"
    direction="row"
    aria-hidden="true"
  >
    <div
      class="StyledTextInput__StyledTextInputContainer-bPTWQl hasTun"
    >
      <input
        type="text"
        class="StyledTextInput-bzOzsW VzRqL"
        id="select"
        autocomplete="off"
        style="cursor: pointer;"
        tabindex="-1"
        readonly=""
      />
    </div>
    <div
      class="StyledBox-bStriS jagiFv"
      direction="column"
      style="min-width: auto;"
    >
      <svg
        class="StyledIcon-gTfjEw htxeRg"
        width="24px"
        height="24px"
        viewBox="0 0 24 24"
        version="1.1"
        role="img"
        aria-label="FormDown"
      >
        <polyline
          fill="none"
          stroke="#000"
          stroke-width="2"
          points="18 9 12 15 6 9"
        />
      </svg>
    </div>
  </div>
</button>
`;

exports[`Select mounts 2`] = `
<button
  class="StyledButton-gXzblK cfJbKA"
  aria-label="undefined"
  type="button"
>
  <div
    class="StyledBox-bStriS janhJy"
    direction="row"
    aria-hidden="true"
  >
    <div
      class="StyledTextInput__StyledTextInputContainer-bPTWQl hasTun"
    >
      <input
        type="text"
        class="StyledTextInput-bzOzsW VzRqL"
        id="select"
        autocomplete="off"
        style="cursor: pointer;"
        tabindex="-1"
        readonly=""
      />
    </div>
    <div
      class="StyledBox-bStriS jagiFv"
      direction="column"
      style="min-width: auto;"
    >
      <svg
        class="StyledIcon-gTfjEw htxeRg"
        width="24px"
        height="24px"
        viewBox="0 0 24 24"
        version="1.1"
        role="img"
        aria-label="FormDown"
      >
        <polyline
          fill="none"
          stroke="#000"
          stroke-width="2"
          points="18 9 12 15 6 9"
        />
      </svg>
    </div>
  </div>
</button>
`;

exports[`Select mounts 3`] = `
<div
  class="StyledBox-bStriS iQipYn"
  direction="column"
  id="select-drop__select"
>
  <div
    class="StyledBox-bStriS feBFzz"
    direction="column"
    overflow="auto"
  >
    <div
      class="StyledBox-bStriS jDxevM"
      direction="column"
      role="menubar"
      tabindex="-1"
    >
      <button
        class="StyledButton-gXzblK khQPBH"
        role="menuitem"
        type="button"
      >
        <div
          class="StyledBox-bStriS gedxom"
          direction="column"
        >
          <span
            class="StyledText-eJMSZI bUVmvh"
          >
            one
          </span>
        </div>
      </button>
      <button
        class="StyledButton-gXzblK khQPBH"
        role="menuitem"
        type="button"
      >
        <div
          class="StyledBox-bStriS gedxom"
          direction="column"
        >
          <span
            class="StyledText-eJMSZI bUVmvh"
          >
            two
          </span>
        </div>
      </button>
    </div>
  </div>
</div>
`;

exports[`Select mounts 4`] = `
"

.iQipYn {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 0;
  padding: 0;
}

@media only screen and (min-width:700px) {
  .cfJbKA {
    -webkit-transition: 0.1s ease-in-out;
    transition: 0.1s ease-in-out;
  }
}

@media only screen and (min-width:700px) {
  .khQPBH {
    -webkit-transition: 0.1s ease-in-out;
    transition: 0.1s ease-in-out;
  }
}"
`;

exports[`Select mounts 5`] = `
<div
  class="StyledBox-bStriS jDxevM"
  direction="column"
  role="menubar"
  tabindex="-1"
>
  <button
    class="StyledButton-gXzblK khQPBH"
    role="menuitem"
    type="button"
  >
    <div
      class="StyledBox-bStriS gedxom"
      direction="column"
    >
      <span
        class="StyledText-eJMSZI bUVmvh"
      >
        one
      </span>
    </div>
  </button>
  <button
    class="StyledButton-gXzblK khQPBH"
    role="menuitem"
    type="button"
  >
    <div
      class="StyledBox-bStriS gedxom"
      direction="column"
    >
      <span
        class="StyledText-eJMSZI bUVmvh"
      >
        two
      </span>
    </div>
  </button>
</div>
`;

exports[`Select mounts with complex options and children 1`] = `
<button
  class="StyledButton-gXzblK cfJbKA"
  aria-label="undefined"
  type="button"
>
  <div
    class="StyledBox-bStriS janhJy"
    direction="row"
    aria-hidden="true"
  >
    <div
      class="StyledTextInput__StyledTextInputContainer-bPTWQl hasTun"
    >
      <input
        type="text"
        class="StyledTextInput-bzOzsW VzRqL"
        id="select"
        autocomplete="off"
        style="cursor: pointer;"
        tabindex="-1"
        readonly=""
      />
    </div>
    <div
      class="StyledBox-bStriS jagiFv"
      direction="column"
      style="min-width: auto;"
    >
      <svg
        class="StyledIcon-gTfjEw htxeRg"
        width="24px"
        height="24px"
        viewBox="0 0 24 24"
        version="1.1"
        role="img"
        aria-label="FormDown"
      >
        <polyline
          fill="none"
          stroke="#000"
          stroke-width="2"
          points="18 9 12 15 6 9"
        />
      </svg>
    </div>
  </div>
</button>
`;

exports[`Select mounts with complex options and children 2`] = `
<button
  class="StyledButton-gXzblK cfJbKA"
  aria-label="undefined"
  type="button"
>
  <div
    class="StyledBox-bStriS janhJy"
    direction="row"
    aria-hidden="true"
  >
    <div
      class="StyledTextInput__StyledTextInputContainer-bPTWQl hasTun"
    >
      <input
        type="text"
        class="StyledTextInput-bzOzsW VzRqL"
        id="select"
        autocomplete="off"
        style="cursor: pointer;"
        tabindex="-1"
        readonly=""
      />
    </div>
    <div
      class="StyledBox-bStriS jagiFv"
      direction="column"
      style="min-width: auto;"
    >
      <svg
        class="StyledIcon-gTfjEw htxeRg"
        width="24px"
        height="24px"
        viewBox="0 0 24 24"
        version="1.1"
        role="img"
        aria-label="FormDown"
      >
        <polyline
          fill="none"
          stroke="#000"
          stroke-width="2"
          points="18 9 12 15 6 9"
        />
      </svg>
    </div>
  </div>
</button>
`;

exports[`Select mounts with complex options and children 3`] = `
<div
  class="StyledBox-bStriS iQipYn"
  direction="column"
  id="select-drop__select"
>
  <div
    class="StyledBox-bStriS feBFzz"
    direction="column"
    overflow="auto"
  >
    <div
      class="StyledBox-bStriS jDxevM"
      direction="column"
      role="menubar"
      tabindex="-1"
    >
      <button
        class="StyledButton-gXzblK khQPBH"
        role="menuitem"
        type="button"
      >
        <span>
          one
        </span>
      </button>
      <button
        class="StyledButton-gXzblK khQPBH"
        role="menuitem"
        type="button"
      >
        <span>
          two
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`Select mounts with complex options and children 4`] = `
"

.iQipYn {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 0;
  padding: 0;
}

@media only screen and (min-width:700px) {
  .cfJbKA {
    -webkit-transition: 0.1s ease-in-out;
    transition: 0.1s ease-in-out;
  }
}

@media only screen and (min-width:700px) {
  .khQPBH {
    -webkit-transition: 0.1s ease-in-out;
    transition: 0.1s ease-in-out;
  }
}"
`;

exports[`Select mounts with search 1`] = `
<div
  class="StyledBox-bStriS iQipYn"
  direction="column"
  id="select-drop__select"
>
  <div
    class="StyledBox-bStriS fGEMtx"
    direction="column"
  >
    <div
      class="StyledTextInput__StyledTextInputContainer-bPTWQl hasTun"
    >
      <input
        type="search"
        class="StyledTextInput-bzOzsW bNmXcC"
        autocomplete="off"
        value=""
      />
    </div>
  </div>
  <div
    class="StyledBox-bStriS feBFzz"
    direction="column"
    overflow="auto"
  >
    <div
      class="StyledBox-bStriS jDxevM"
      direction="column"
      role="menubar"
      tabindex="-1"
    >
      <button
        class="StyledButton-gXzblK khQPBH"
        role="menuitem"
        type="button"
      >
        <div
          class="StyledBox-bStriS gedxom"
          direction="column"
        >
          <span
            class="StyledText-eJMSZI bUVmvh"
          >
            one
          </span>
        </div>
      </button>
      <button
        class="StyledButton-gXzblK khQPBH"
        role="menuitem"
        type="button"
      >
        <div
          class="StyledBox-bStriS gedxom"
          direction="column"
        >
          <span
            class="StyledText-eJMSZI bUVmvh"
          >
            two
          </span>
        </div>
      </button>
    </div>
  </div>
</div>
`;

exports[`Select mounts with search 2`] = `
"

.iQipYn {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 0;
  padding: 0;
}

@media only screen and (min-width:700px) {
  .cfJbKA {
    -webkit-transition: 0.1s ease-in-out;
    transition: 0.1s ease-in-out;
  }
}

@media only screen and (min-width:700px) {
  .khQPBH {
    -webkit-transition: 0.1s ease-in-out;
    transition: 0.1s ease-in-out;
  }
}"
`;

exports[`Select mounts with search 3`] = `
<input
  type="search"
  class="StyledTextInput-bzOzsW bNmXcC"
  autocomplete="off"
  value=""
/>
`;

exports[`Select mounts with search 4`] = `
<div
  class="StyledBox-bStriS iQipYn"
  direction="column"
  id="select-drop__select"
>
  <div
    class="StyledBox-bStriS fGEMtx"
    direction="column"
  >
    <div
      class="StyledTextInput__StyledTextInputContainer-bPTWQl hasTun"
    >
      <input
        type="search"
        class="StyledTextInput-bzOzsW bNmXcC"
        autocomplete="off"
        value="a"
      />
    </div>
  </div>
  <div
    class="StyledBox-bStriS feBFzz"
    direction="column"
    overflow="auto"
  >
    <div
      class="StyledBox-bStriS jDxevM"
      direction="column"
      role="menubar"
      tabindex="-1"
    >
      <button
        class="StyledButton-gXzblK khQPBH"
        role="menuitem"
        type="button"
      >
        <div
          class="StyledBox-bStriS gedxom"
          direction="column"
        >
          <span
            class="StyledText-eJMSZI bUVmvh"
          >
            one
          </span>
        </div>
      </button>
      <button
        class="StyledButton-gXzblK khQPBH"
        role="menuitem"
        type="button"
      >
        <div
          class="StyledBox-bStriS gedxom"
          direction="column"
        >
          <span
            class="StyledText-eJMSZI bUVmvh"
          >
            two
          </span>
        </div>
      </button>
    </div>
  </div>
</div>
`;

exports[`Select mounts with search 5`] = `
"

.iQipYn {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 0;
  padding: 0;
}

@media only screen and (min-width:700px) {
  .cfJbKA {
    -webkit-transition: 0.1s ease-in-out;
    transition: 0.1s ease-in-out;
  }
}

@media only screen and (min-width:700px) {
  .khQPBH {
    -webkit-transition: 0.1s ease-in-out;
    transition: 0.1s ease-in-out;
  }
}"
`;
